ARCs/arc-0063.md

```diff
@@ -0,0 +1,322 @@
+==
+&&
+txn Fee
+global MinTxnFee
+==
+&&
+return
+"""
+compiled_program = client.compile(teal_program)
+program = base64.b64decode(compiled_program["result"])
+lsig = transaction.LogicSigAccount(program)
+```
+
+> This allow the application with the id `app_client.app_id` to control the signer
+
+To be consistant with the previous example, we will use a similar opt-in process.
+
+```python
+class SmartApp(ARC4Contract):
+    def __init__(self) -> None:
+        self.db = BoxMap(Account, String, key_prefix="")
+
+    @abimethod()
+    def opt_in(self, id: UInt64, account: Account) -> None:
+        itxn.AssetTransfer(
+            asset_amount=0,
+            xfer_asset=id,
+            sender=account,
+            asset_receiver=account,
+            fee=1000,
+        ).submit()
+
+    @abimethod()
+    def set_public_sig(self, account: Account, sig: String) -> bool:
+        self.db[account] = sig
+        return self.db[account] == sig
+
+    @abimethod(readonly=True)
+    def get_public_sig(self, account: Account) -> String:
+        return self.db[account]
+```
+
+1. **Generate Plug-In Signer**:
+    - Generate a random new account that we will only use once.
+
+    ```python
+    plug_in_sk, plug_in_addr  = account.generate_account()
+    ```
+
+2. **Sign Lsig with Plug-In Signer**:
+    - Sign the Lsig using the plug-in signer.
+
+    ```python
+        public_key, secret_key = nacl.bindings.crypto_sign_seed_keypair(base64.b64decode(plug_in_sk)[: constants.key_len_bytes])
+        message = constants.logic_prefix + program
+        raw_signed = nacl.bindings.crypto_sign(message, secret_key)
+        crypto_sign_BYTES = nacl.bindings.crypto_sign_BYTES
+        signature = nacl.encoding.RawEncoder.encode(raw_signed[:crypto_sign_BYTES])
+        plug_in_public_sig = base64.b64encode(signature).decode()
+    ```
+
+3. **Create 1/2 Msig Account**:
+    - Create a multi-signature account with owner address, the Lsig address, and the plug-in signer.
+
+    ```python
+    owner_vault_msig = transaction.Multisig(1,1,[owner_addr, plug_in_addr])
+    ```
+
+    - Publish the public signature by using set_public_sig to the app.
+
+    ```python
+    response = app_client.set_public_sig(
+        account=owner_addr,
+        sig=plug_in_public_sig,
+        transaction_parameters=algokit_utils.TransactionParameters(
+            boxes=[(app_client.app_id, encoding.decode_address(owner_addr))],
+            accounts=[owner_addr]
+        )
+    )
+    ```
+
+    - Get the transaction set_public_sig from the app.
+
+    ```python
+    response = app_client.get_public_sig(
+        account=owner_addr,
+        transaction_parameters=algokit_utils.TransactionParameters(
+            boxes=[(app_client.app_id, encoding.decode_address(owner_addr))]
+        ),
+    )
+    ```
+
+4. **App in control of Msig Vault**:
+    - Anyone can now call the app to opt-in any asset to the Msig vault using the plug-in signer’s public address and the published signature.
```

Comment:
I'm not sure I understand this section. The msig account hasn't been rekeyed to the application, so how does the application control the multisig vault at this point? All it has is a list of addresses to signatures, which haven't been verified to be accurate, presumably because it's an early design? Or have I misunderstood it, and this application is just there to provide a mapping service between address and signature?

Reply:
With the signature `plug_in_public_sig` of the `plug_in_addr` that is now stored in the app thanks to the `set_public_sig()` method, anyone can get this value by looking at the box/ calling the `get_public_sig()` method. And recontruct a multisignature to execute the application call. 
By executing the call bellow. 

> TLDR `plug_in_addr` signed the `teal_program` and it is enough to allow an inner transaction doing operation on the msig account 

```json
{
    ...
    inner-txns:[],
    signature:{
    logicsig:{
    logic:"CjEQgQYSMRiB7QcSEDEgMgMSEDEBMgASEEM=",
    multisig-signature:{
        version:1,
        threshold:1,
        subsignature:[
                {...},
                {
                    public-key:"24RB9iNUHfFDelIYDnw9URA2Pf6TYz8vg/vom9sUwAI=",
                    signature:"iPu354yel81lxlJTlFPG34AxDVR5Bc1MLL1zRwWwK7zM2p4Ehkf5ktOH2H67avgO31a0v8ejr/DL328jNAj4BA=="
                }
            ]
        }
    }
}
```